cpp/src/strings/split/split_record.cu

```diff
@@ -23,123 +25,56 @@
 
-using string_index_pair = thrust::pair<const char*, size_type>;
-
 namespace {
 
-enum class Dir { FORWARD, BACKWARD };
-
-/**
- * @brief Compute the number of tokens for the `idx'th` string element of `d_strings`.
- *
- * The number of tokens is the same regardless if counting from the beginning
- * or the end of the string.
- */
-struct token_counter_fn {
-  column_device_view const d_strings;  // strings to split
-  string_view const d_delimiter;       // delimiter for split
-  size_type const max_tokens = std::numeric_limits<size_type>::max();
-
-  __device__ size_type operator()(size_type idx) const
-  {
-    if (d_strings.is_null(idx)) { return 0; }
-
-    auto const d_str      = d_strings.element<string_view>(idx);
-    size_type token_count = 0;
-    size_type start_pos   = 0;
-    while (token_count < max_tokens - 1) {
-      auto const delimiter_pos = d_str.find(d_delimiter, start_pos);
-      if (delimiter_pos == string_view::npos) break;
-      token_count++;
-      start_pos = delimiter_pos + d_delimiter.length();
-    }
-    return token_count + 1;  // always at least one token
-  }
-};
-
-/**
- * @brief Identify the tokens from the `idx'th` string element of `d_strings`.
- */
-template <Dir dir>
-struct token_reader_fn {
-  column_device_view const d_strings;  // strings to split
-  string_view const d_delimiter;       // delimiter for split
-  int32_t* d_token_offsets{};          // for locating tokens in d_tokens
-  string_index_pair* d_tokens{};
-
-  __device__ string_index_pair resolve_token(string_view const& d_str,
-                                             size_type start_pos,
-                                             size_type end_pos,
-                                             size_type delimiter_pos) const
-  {
-    if (dir == Dir::FORWARD) {
-      auto const byte_offset = d_str.byte_offset(start_pos);
-      return string_index_pair{d_str.data() + byte_offset,
-                               d_str.byte_offset(delimiter_pos) - byte_offset};
-    } else {
-      auto const byte_offset = d_str.byte_offset(delimiter_pos + d_delimiter.length());
-      return string_index_pair{d_str.data() + byte_offset,
-                               d_str.byte_offset(end_pos) - byte_offset};
-    }
+template <typename Tokenizer>
+std::unique_ptr<column> split_record_fn(strings_column_view const& input,
+                                        Tokenizer tokenizer,
+                                        rmm::cuda_stream_view stream,
+                                        rmm::mr::device_memory_resource* mr)
+{
+  if (input.is_empty()) { return make_empty_column(type_id::LIST); }
+  if (input.size() == input.null_count()) {
+    auto offsets = std::make_unique<column>(input.offsets(), stream, mr);
+    auto results = std::make_unique<column>(input.parent(), stream, mr);
+    return make_lists_column(input.size(),
+                             std::move(offsets),
+                             std::move(results),
```

Comment:
Just to make sure I understand correctly? The output lists column will be lists of strings (all nulls)?

Reply:
I ask this because if we want to have an output lists column of all null, we just need to have an empty child, instead of all-nulls child.